flang/lib/Lower/OpenMP.cpp

```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
       }
     }
     createBodyOfOp<omp::ParallelOp>(parallelOp, firOpBuilder, currentLocation);
-  } else if (blockDirective.v == llvm::omp::OMPD_master) {
+}
+
+static void
+genOMP(Fortran::lower::AbstractConverter &converter,
+       Fortran::lower::pft::Evaluation &eval,
+       const Fortran::parser::OpenMPBlockConstruct &blockConstruct) {
+  const auto &beginBlockDirective =
+      std::get<Fortran::parser::OmpBeginBlockDirective>(blockConstruct.t);
+  const auto &blockDirective =
+      std::get<Fortran::parser::OmpBlockDirective>(beginBlockDirective.t);
+
+  if (blockDirective.v == llvm::omp::OMPD_parallel)
+    createParallelOp<Fortran::parser::OmpBeginBlockDirective>(
+        converter, eval,
+        std::get<Fortran::parser::OmpBeginBlockDirective>(blockConstruct.t));
+  else if (blockDirective.v == llvm::omp::OMPD_master) {
+    llvm::ArrayRef<mlir::Type> argTy;
+    auto &firOpBuilder = converter.getFirOpBuilder();
+    auto currentLocation = converter.getCurrentLocation();
     auto masterOp =
         firOpBuilder.create<mlir::omp::MasterOp>(currentLocation, argTy);
```

Comment:
Is this `MasterOp` ever going to have arguments? (Not sure why this one needs an empty list of argument types.)

Reply:
I don't think it is going to have arguments. Also, the Master construct is deprecated in OpenMP 5.1.
There is a build function without arguments. I don't know whether @SouraVX faced any issues with using it.

```
void MasterOp::build(::mlir::OpBuilder &odsBuilder, ::mlir::OperationState &odsState) {
  (void)odsState.addRegion();
}          
```

Reply:
I've noted down the comment when it was made, but unfortunately didn't paid much attention to it since this patch was my primary priority. I'll take a look at this again, a bit later.

Reply:
OK with me to address in a separate patch.